object_storage_api/repositories/image.py

```diff
@@ -49,13 +50,19 @@ def get(self, image_id: str, session: ClientSession = None) -> Optional[ImageOut
         :param image_id: ID of the image to retrieve.
         :param session: PyMongo ClientSession to use for database operations.
         :return: Retrieved image or `None` if not found.
+        :raises MissingRecordError: If the supplied `image_id` is invalid or non-existent.
         """
-        image_id = CustomObjectId(image_id)
         logger.info("Retrieving image with ID: %s from the database", image_id)
-        image = self._images_collection.find_one({"_id": image_id}, session=session)
+        message = f"Image with image_id {image_id} was not found."
+        entity_name = "image"
+        try:
+            image_id = CustomObjectId(image_id)
+            image = self._images_collection.find_one({"_id": image_id}, session=session)
+        except InvalidObjectIdError as exc:
+            raise MissingRecordError(detail=message, entity_name=entity_name) from exc
```

Comment:
Sorry just had another think about he errors and where they should be raised, I think I have changed my mind from before 😄

Thinking about the repo as the interface to the database – I would expect that doing `repo.get(“invalid-id”)` should raise an invalid id error. As an example, for IMS within `CatalogueCategoryService.create`’s parent check, we use the `repo.get` – if the parent id was actually invalid but we checked existence by catching the MissingRecordError then creating a catalogue category with a `parent_id: “invalid-id”` would succeed with a `parent_id` of `None` when it should fail due to an invalid ID. 

Due to this I actually think that this part needs to be in the service instead 🤦‍♂️

Reply:
To differentiate between an invalid ID and missing record we could look at the `detail `of the `MissingRecordError` in the service layer. Implementing something like this, whilst keeping most of the error handling in the repo layer.

```
parent_id = catalogue_category.parent_id
        try:
            parent_catalogue_category = self.get(parent_id) if parent_id else None
        except MissingRecordError as exc:
            if exc.detail == f"Invalid catalogue category ID: {parent_id}":
                raise exc
            elif exc.detail == f"Missing catalogue category: {parent_id}":
                
```

But I do agree, that from a "layer's purpose" perspective, the repo layer should just be throwing errors related to the database, whilst the service layer takes care of more of the "business logic".

Reply:
Actually, I've been thinking about it more and I have another idea.

In the repo layer, we could raise a `MissingRecordError` and `InvalidObjectIDerror` respectively for the 2 separate error cases. We could change the `status_code `of the `InvalidObjectIDerror` when its caught in the repo layer to a `404`, and let it raise as normal. In the service layer this makes it easy to differentiate it, and when the response reaches the user, it just seems like the entity didn't exist since the status code and response detail is set to `404 `and `Entity not found` respectively.

Reply:
That sounds perfect, the value logged would still show it was due to the invalid id but when raised it would still look the same. Good idea!